### PR TITLE
[AndroidCrypto] Enable using X.509 certs with DSA

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Hook up the DSAAndroid implementation to the X.509 layer

### CODE REVIEW

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c [hunk 1]
```diff
@@ -260,7 +260,7 @@ void* AndroidCryptoNative_X509PublicKey(jobject /*X509Certificate*/ cert, PAL_Ke
             keyHandle = AndroidCryptoNative_NewEcKeyFromPublicKey(env, key);
             break;
         case PAL_DSA:
-            keyHandle = NULL;
+            keyHandle = (*env)->NewObject(env, g_keyPairClass, g_keyPairCtor, key, NULL);
```
Comment (bartonjs) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:263]: I won't say that this is incorrect... but I'll say that it stands out as notably different from the above and below case blocks.

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/AndroidCertificatePal.cs [hunk 2]
```diff
@@ -386,7 +393,22 @@ internal void SetPrivateKey(SafeKeyHandle privateKey)
...
+            if (typedKey != null)
+            {
+                return CopyWithPrivateKeyHandle(typedKey.DuplicateKeyHandle());
+            }
+
+            DSAParameters dsaParameters = privateKey.ExportParameters(true);
+            using (PinAndClear.Track(dsaParameters.G!))
+            using (PinAndClear.Track(dsaParameters.P!))
+            using (PinAndClear.Track(dsaParameters.Q!))
+            using (PinAndClear.Track(dsaParameters.X!))
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/AndroidCertificatePal.cs:406]: P, G, and Q are public parameters, they don't need to use PinAndClear.Track.  (Only X does)

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:263]: Yes, it does stand out. I'll add an abstraction layer.

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/X509Pal.cs [hunk 3]
```diff
@@ -130,6 +140,41 @@ private static RSA DecodeRsaPublicKey(byte[] encodedKeyValue)
...
+                AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+                spki.Encode(writer);
+
+                byte[] rented = CryptoPool.Rent(writer.GetEncodedLength());
+
+                if (!writer.TryEncode(rented, out int written))
+                {
+                    Debug.Fail("TryEncode failed with a pre-allocated buffer");
+                    throw new InvalidOperationException();
+                }
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/X509Pal.cs:161]: This boiler-plate can be written easier:

```suggestion
                int written = writer.Encode(rented);
```

It's just an "older code vs new API" artifact that there are so many blocks like this through the codebase.

File: src/libraries/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj [hunk 4]
```diff
@@ -444,6 +446,8 @@
              Link="Common\System\Security\Cryptography\RSAAndroid.cs" />
     <Compile Include="$(CommonPath)System\Security\Cryptography\RsaPaddingProcessor.cs"
              Link="Common\System\Security\Cryptography\RsaPaddingProcessor.cs" />
+    <Compile Include="$(CommonPath)System\Security\Cryptography\DSAAndroid.cs"
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj:449]: Move above ECAndroid for alpha ordering / ease of finding.

**Comment by bartonjs** [comment]
Assuming that the tests look like they're generally (ideally "all") passing: LGTM with minor feedback.

**Review by bartonjs — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Wasm failure is unrelated.

**Pull request merged**